drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua

```diff
@@ -0,0 +1,1580 @@
+end
+
+----------------------------------
+-- Number Of PIN User Supported --
+----------------------------------
+local function pin_users_supported_handler(driver, device, ib, response)
+  device:emit_event(capabilities.lockCredentials.pinUsersSupported(ib.data.value, {visibility = {displayed = false}}))
+end
+
+-------------------------
+-- Min PIN Code Length --
+-------------------------
+local function min_pin_code_len_handler(driver, device, ib, response)
+  device:emit_event(capabilities.lockCredentials.minPinCodeLen(ib.data.value, {visibility = {displayed = false}}))
+end
+
+-------------------------
+-- Max PIN Code Length --
+-------------------------
+local function max_pin_code_len_handler(driver, device, ib, response)
+  device:emit_event(capabilities.lockCredentials.maxPinCodeLen(ib.data.value, {visibility = {displayed = false}}))
+end
+
+--------------------------------------
+-- Require PIN For Remote Operation --
+--------------------------------------
+--- If a device needs a cota credential this function attempts to set the credential
+--- at the index provided. The set_credential_response_handler handles all failures
+--- and retries with the appropriate index when necessary.
+local function set_cota_credential(device, credential_index)
+  local eps = device:get_endpoints(DoorLock.ID)
+  local cota_cred = device:get_field(lock_utils.COTA_CRED)
+  if cota_cred == nil then
+    -- Shouldn't happen but defensive to try to figure out if we need the cota cred and set it.
+    device:send(DoorLock.attributes.RequirePINforRemoteOperation:read(device, #eps > 0 and eps[1] or 1))
+    device.thread:call_with_delay(2, function(t) set_cota_credential(device, credential_index) end)
+  elseif not cota_cred then
+    device.log.debug("Device does not require PIN for remote operation. Not setting COTA credential")
+    return
+  end
+
+  if device:get_field(lock_utils.SET_CREDENTIAL) ~= nil then
+    device.log.debug("delaying setting COTA credential since a credential is currently being set")
+    device.thread:call_with_delay(2, function(t)
+      set_cota_credential(device, credential_index)
+    end)
+    return
+  end
+
+  device:set_field(lock_utils.COTA_CRED_INDEX, credential_index, {persist = true})
+  local credential = {credential_type = DoorLock.types.CredentialTypeEnum.PIN, credential_index = credential_index}
+  -- Set the credential to a code
+  device:set_field(lock_utils.BUSY_STATE, true, {persist = true})
+  device:set_field(lock_utils.COMMAND_NAME, "addCota")
+  device:set_field(lock_utils.CRED_INDEX, credential_index)
+  device:set_field(lock_utils.SET_CREDENTIAL, credential_index)
+  device:set_field(lock_utils.USER_TYPE, "remote")
+  device.log.info(string.format("Attempting to set COTA credential at index %s", credential_index))
+  device:send(DoorLock.server.commands.SetCredential(
+    device,
+    #eps > 0 and eps[1] or 1,
```

Comment:
Should `eps` be sorted before this like we do in `find_default_endpoint`?

Reply:
If the device has two or more lock clusters, it seems that we need to find an endpoint that supports COTA. However, I think it will be okay as it is because there is no possibility that such a device will come out in my personal opinion.